### ISSUE TITLE
Turn JsonApiDocument into an enum for Success, Failure and Info

### ISSUE THREAD

**Issue by michiel** [michiel/jsonapi-rust#24]
As suggested in #23, turn `JsonApiDocument` into an enum and see how that changes the flow of the code.

**Comment by scotttrinh** [comment]
I'm implementing a slight variation of this library as a JSON:API wrapper in my own library and have done this. My "top-level document" type looks like:

```rust
pub enum Document {
    Meta {
        meta: MetaValue,
    },

    Error {
        errors: Vec<ApiError>,
        #[serde(skip_serializing_if = "Option::is_none")]
        links: Option<LinksValue>,
        #[serde(skip_serializing_if = "Option::is_none")]
        meta: Option<MetaValue>,
    },

    Data {
        data: ApiResource,
        #[serde(skip_serializing_if = "Option::is_none")]
        links: Option<LinksValue>,
        #[serde(skip_serializing_if = "Option::is_none")]
        meta: Option<MetaValue>,
        #[serde(skip_serializing_if = "Option::is_none")]
        included: Option<IncludedValue>,
    },
}
```

Haven't done many things with it yet, just a single simple response type, so I can't speak to the ergonomics pros and cons, but keeps you from having to do run-time validation on something that seems like the compiler can verify for us. I'd be interested in helping to implement this here, but it's definitely a breaking change.

**Comment by JadedEvan** [comment]
@scotttrinh I think your concept is headed in the right direction. I have been exploring the tests in this repo and discovered that aspects of the current implementation are not compliant. Currently this crate does not respect some of the top-level document guidelines

- A document MUST contain at least one of the following top-level members:
- The members data and errors MUST NOT coexist in the same document.

I think the `enum` is the right concept, but may also lack the rules outlined by the spec. I'm going to explore this concept more - as I do agree that it would be valuable to let Rust's type system handle some of these validations

**Comment by JadedEvan** [comment]
@scotttrinh thanks again for your suggestion - I went ahead and implemented this change in #75 which was just merged. We haven't cut a versioned release for this yet but we'll be pushing that out in the next few days.
